### PR TITLE
[FIX] mail: Incorrect tracking values

### PR DESCRIPTION
**Pull request by sauc-odoo**
Before this commit user was getting incorrect values in the statusbar when user is changing stage 
from one stage to another. The duration of previous stage is carried forward to the next stage.

Steps to reproduce:
-  ​Open a demo data lead or task and move it to another stage from statusbar in form view.

Observed behavior:
​The tracking values in the statusbar are incorrect.

Expected behavior:
​When moving to a new stage for example, it should not keep the old value.

After this commit user will be able to get correct values in the statusbar when user will change stage 
from one stage to another.

Task-3584456

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/mail/tests/mail_tracking_duration_mixin_case.py [hunk 1]
```diff
@@ -195,5 +195,5 @@ def _test_queries_batch_duration_tracking(self):
         self.flush_tracking()
         batch[self.track_duration_field] = self.stage_2.id
 
-        with self.assertQueryCount(2):
+        with self.assertQueryCount(4):
```
Comment (Odoonan) [inline addons/mail/tests/mail_tracking_duration_mixin_case.py:198]: If the issue did not break any test, maybe we could add a case (ideally, in one of the existing tests) to be sure it does not occur again in the future? :)

File: addons/mail/static/src/views/fields/statusbar_duration/statusbar_duration_field.js [hunk 2]
```diff
@@ -13,7 +13,10 @@ export class StatusBarDurationField extends StatusBarField {
         const durationTracking = this.props.record.data.duration_tracking || {};
         if (Object.keys(durationTracking).length) {
             for (const item of items) {
-                const duration = durationTracking[item.value];
+                let duration = durationTracking[item.value];
+                if (duration == 0) {
+                    duration = 1;
+                }
```
Comment (Odoonan) [inline addons/mail/static/src/views/fields/statusbar_duration/statusbar_duration_field.js:19]:
```
let duration = durationTracking[item.value];
duration = duration ? duration : 1;
```
but then, do we not alway enter the if below ? Could we have negative values ?

File: addons/mail/models/mail_tracking_duration_mixin.py [hunk 3]
```diff
@@ -56,8 +56,9 @@ class MyModel(models.Model):
                   AND m.res_id IN %(record_ids)s
              ORDER BY v.id
         """
-        self.env.cr.execute(query, {"field_id": field.id, "model_name": self._name, "record_ids": tuple(self.ids)})
-        trackings = self.env.cr.dictfetchall()
+        with self.env.cr.savepoint():
```
Comment (Odoonan) [inline addons/mail/models/mail_tracking_duration_mixin.py:59]: Can you please explain the role of this in your fix? :)

**Review by Odoonan — COMMENTED** [review]
Hello @sauc-odoo 

Thanks for your work, good catch.

I've already left a few remarks / questions, just to be sure to understand the issue correctly.

Cheers!

File: addons/mail/models/mail_tracking_duration_mixin.py (hunk 3 above)
Comment (sauc-odoo) [reply addons/mail/models/mail_tracking_duration_mixin.py:59]: It is used to create a savepoint within the current database transaction. A savepoint is a point in a transaction where we can roll back if needed. 
By using this in our case we will get the values of all the previous stages in `trackings` and when function calls again all the values saved in savepoint will execute. Previously the values in `trackings` were lost when function calls again.

So to not keep old value in the next stage in statusbar i have implied this.

File: addons/mail/static/src/views/fields/statusbar_duration/statusbar_duration_field.js (hunk 2 above)
Comment (sauc-odoo) [reply addons/mail/static/src/views/fields/statusbar_duration/statusbar_duration_field.js:19]: Yes, we will always enter the if over here. I have added this code to get `1m` as soon as i click on next stage. If i don't add it we will not get any time on next stage as soon as i click on it.
We can also remove it if we don't want `1m` as soon as i click on next stage.

I will change my code and add the ternary operator as you have suggested.
Thanks!!

File: addons/mail/tests/mail_tracking_duration_mixin_case.py (hunk 1 above)
Comment (sauc-odoo) [reply addons/mail/tests/mail_tracking_duration_mixin_case.py:198]: Hello @Odoonan 

The test in which i have made changes was breaking because before my fix only 2 queries were running in that test but after my fix 4 queries will execute. So after the updation of this test case where i have added `4` in `self.assertQueryCount():` will test my fix.

Thanks!!

Comment (Odoonan) [reply addons/mail/tests/mail_tracking_duration_mixin_case.py:198]: Hello, I see. I meant that maybe we could add a small test case in one of the existing tests to ensure that the issue you fix do not break again later. (So a small test case that would break without your fix but not with it)

Cheers!

File: addons/mail/static/src/views/fields/statusbar_duration/statusbar_duration_field.js (hunk 2 above)
Comment (sauc-odoo) [reply addons/mail/static/src/views/fields/statusbar_duration/statusbar_duration_field.js]: For above we are having 3 cases - 

1. For the upcoming stages : we will get duration null so it will go to else and `item.shortTimeInStage will be set to 0` so that we does not get any tracking of it.

2. For previous stages or current stage : we will get duration a positive value so it will go in if(duration>0) and `item.shortTimeInStage = formatDuration(duration, false);` so that we will get tracking of it.

3. For next stage on which we are clicking : we will get duration=0 so it will also go in else and `item.shortTimeInStage will be set to 0` so that we does not get any tracking of it.

But after the fix - 
```
let duration = durationTracking[item.value];
duration = (duration==0) ? 1 : duration;
```

First two case will remain same but for 3 case - we will get duration=0 which will be converted to 1 and then it will go in if(duration>0) and `item.shortTimeInStage = formatDuration(duration, false);` so that we will get tracking of it as `1m`.

I hope it might help you to understand. :slightly_smiling_face:

File: addons/mail/tests/mail_tracking_duration_mixin_case.py (hunk 1 above)
Comment (sauc-odoo) [reply addons/mail/tests/mail_tracking_duration_mixin_case.py:198]: Hello @Odoonan 
If i remove my fix i will get 2 queries executed in above test case which will not fail because in `assertQueryCount()` if the count is less than expected (i.e. 4 in our case) a logger info is generated from here - 
https://github.com/odoo/odoo/blob/ee31c84c360425b0acba4f04462bc752c121b94b/odoo/tests/common.py#L541-L544

So to add a case that would break without my fix but not with it i have to create a method `assertQueryCountEqual()` similar to `assertQueryCount()` in which the count should be equal to expected otherwise it will fail. So can i add a method like it in `odoo/community/odoo/tests/common.py`, Kindly suggest me!!

Comment (Odoonan) [reply addons/mail/tests/mail_tracking_duration_mixin_case.py:198]: Hello @sauc-odoo . I think there is a misunderstanding here, let me rephrase it. The `self.assertQueryCount(4)` is fine. You can keep it as indeed, there are more queries.

What I meant was to **also** add a unit test for the fix itself. For instance

1. create a record in a stage and 'wait' 5 minutes
2. Assert stage duration is set to 5 min
3. Change stage
4. Make sure the new duration is not 5 min

Something like that, do you see what I mean ? But actually it looks like such tests exist, so it is strange that they do not fail.

Eventually, do not update the code yet, I would like to have a deeper look first. :)

Cheers!

File: addons/mail/static/src/views/fields/statusbar_duration/statusbar_duration_field.js [hunk 4]
```diff
@@ -13,7 +13,8 @@ export class StatusBarDurationField extends StatusBarField {
         const durationTracking = this.props.record.data.duration_tracking || {};
         if (Object.keys(durationTracking).length) {
             for (const item of items) {
-                const duration = durationTracking[item.value];
+                let duration = durationTracking[item.value];
+                duration = (duration==0) ? 1 : duration;
```
Comment (Odoonan) [inline addons/mail/static/src/views/fields/statusbar_duration/statusbar_duration_field.js:17]: Actually I see in formatDuration this :

```js
    if (seconds < 60) {
        seconds = 60;
    }
```

So we could just keep old code here but just use `if (duration >= 0)` instead of `if (duration > 0)` :)

Comment (sauc-odoo) [reply addons/mail/static/src/views/fields/statusbar_duration/statusbar_duration_field.js]: Updated the code as per your suggestions. 
Thanks!!

File: addons/mail/models/mail_tracking_duration_mixin.py (hunk 3 above)
Comment (Odoonan) [reply addons/mail/models/mail_tracking_duration_mixin.py:59]: *Small summary*

So here is what I understand of the situation, and how it is leading to the issue. The compute and `_get_duration_from_tracking` method work as expected. However, it computes the durations based on existing records of mail.tracking.value. (+1 fake for current stage)

The issue is that when, on the form view, one changes the stage, the _compute will NOT fetch the lastly created mail.tracking.value (the one tracking the stage change) in the dictfetchall. The weird duration jump from one stage to the other comes from there.

However, if you leave the form view, and go back to it, it will be correctly handled. So the issue is that the _compute would need to have updated values when staying on the same form view.

The savepoint does the trick but feels arbitrary, as what we need is not previous values, but all values including the last one, actually. I would have thought flushes would be enough. I'm still thinking about it.

Comment (sauc-odoo) [reply addons/mail/models/mail_tracking_duration_mixin.py:59]: Hello @Odoonan,

Flushes seems to be working, so should I replace savepoint with flush. Kindly give your suggestions.
Thanks!!

Comment (Odoonan) [reply addons/mail/models/mail_tracking_duration_mixin.py:59]: Hello @sauc-odoo , yes you can update so that I see what you did :)

We can always go back to this later.

Cheers!

Comment (sauc-odoo) [reply addons/mail/models/mail_tracking_duration_mixin.py]: Updated the code as per the suggestions, kindly check it.
Thanks!!

**Comment by tde-banana-odoo** [comment]
See https://github.com/odoo/odoo/pull/161688/files for a more complete fix and discussion
